modules/openapi-generator/src/main/resources/JavaSpring/nullableDataType.mustache

```diff
@@ -1 +1 @@
-{{#isNullable}}JsonNullable<{{{datatypeWithEnum}}}>{{/isNullable}}{{^isNullable}}{{{datatypeWithEnum}}}{{/isNullable}}
\ No newline at end of file
+{{#isNullable}}JsonNullable<{{{datatypeWithEnum}}}>{{/isNullable}}{{^isNullable}}{{>optionalDataType}}{{/isNullable}}
```

Comment:
It seems weird to have both JsonNullable and Optional...

Reply:
I was not sure if the JsonNullable could be removed due to breaking changes.
What do you mean I should remove them?

Reply:
No you can't remove JsonNullable. JsonNullable already has a notion of presence/absence so it's weird to have a double notion for this.
Note that having Optional POJO fields is a bad practice in the first place...

Reply:
> Note that having Optional POJO fields is a bad practice in the first place...

Sorry no it's not really a bad practice. Nullable objects are a bad practice. Since jackson have a JDK8Module, is no reason why you should not use the Optional instead of own JsonNullable object. The Optional works fine.
Here is a good presentation why Optionals are not a bad practice.
https://slides.codefx.org/expert-java-8/2017-09-18-objektforum-stuttgart/index.html#/

I think nullable collections and the JsonNullable is more a code smell and bad practice

Reply:
Exactly in the [presentation you linked](https://slides.codefx.org/expert-java-8/2017-09-18-objektforum-stuttgart/index.html#/_basic_rules): 

> make everyone setting Optional to null buy a round of drinks or wear a silly hat

And that's exactly what you do by setting your field to Optional in the POJO :smiley:  ! :beer: accepted !

Reply:
Ok that's a bug, the default value have to be `Optional.empty()`. We have constructor for that job ;). I will fix that. 

But this was not your reason, why the Optional is a bad practice?

Reply:
Thanks for the hint, fixed it. Now no `Optional` is `null`. Default value is `Optional.empty()`

Reply:
The reasons are all in your presentation (which is really good btw): no serializability, decreased performance due to de-referencing, was designed only as a method return value, etc...
I know there's a lot of bikeshedding over the usage of Optionals so if you want to use them in your POJOs, just do it. Generator options are there for this.

As for JsonNullable, note that it has a different purpose than Optional. It's meant to handle tri-state variables that can happen when doing [JSON Merge Patch](https://tools.ietf.org/html/rfc7386). You need the 3 states: absent, present and null, present and non-null. You can't do that with Optionals.

Reply:
Oh! I see that you want to use Optionals for nullable fields. Don't do that : you can't set an Optional to null so you only have the choice between `empty` and `not-empty` but nullable is about having the choice between `absent`, `null`, and `non-null` so that's different.

Reply:
Yes i like the presentation too. For that reason i would like to use Optional instead of nullable fields. 
Thanks for your hint, i know the tri-state issue, but i thought (i will double check that) there is a solution with Optional. First i thought `empty` `not-empty` and `empty string`, but this will not work with Integer.

Reply:
It looks like you really need `null` as state. Then the difference is not so big except that you do not need the extra lib.
https://github.com/FasterXML/jackson-databind/issues/578
https://github.com/FasterXML/jackson-datatype-jdk8/issues/2

Reply:
You've convinced me. It is really better to use the `JsonNullable`. Optionals and zero cost too much beer ;).

So will close the ticket. Thanks for your support.

One question to the `JsonNullable`. Is there a reason why the class have no method `isUndefined()`?

e.g.

```
public boolean isUndefined(){
return UNDEFINED.equals(this);
}
```